### DISCUSSION TITLE
How to convert four dimensional tetrahedron primitive with normal to 3d triangle primitives with normals

### DISCUSSION THREAD

**Discussion by kijetesantakalu042** [community/community#135196] [Programming Help]
### Body

I'd like to be able to be able to project a 4d tetrahedron with a normal (Something like {{1, 0, 0, 0}. {0, 1, 0, 0}, {0, 0, 1, 0}, {0, 0, 0, 1}, {0, 1, 1, 1}} to a set of 3d triangle primitives. Does anyone know how to do that?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by geniuszly** [comment]
To project a 4D tetrahedron into 3D space, you can use a method similar to perspective projection. The 4D vertices of the tetrahedron can be projected onto a 3D space by "flattening" one of the dimensions. This process is akin to how 3D objects are projected onto a 2D plane. Given your 4D tetrahedron vertices, you can project them into 3D space by ignoring one of the coordinates or by using a perspective projection formula.

Let's assume your 4D vertices are as follows:
```python
vertices_4d = [
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 1, 0],
    [0, 0, 0, 1],
    [0, 1, 1, 1]
]
```

For simplicity, you can project these vertices by dropping one of the coordinates. Alternatively, you can use a perspective projection where the 4th dimension (`w`) is treated as a distance from the 3D "camera"
```python
projected_vertices_3d = [
    [x, y, z] for [x, y, z, w] in vertices_4d
]
```

Once you have the 3D points, you can create the triangles (faces) of the 3D simplex. For a 4-simplex, each face will correspond to a 3-simplex (a tetrahedron), but when projected into 3D, some of these might collapse into triangles. With the 3D vertices and their connectivity (which remains the same as in the original 4D tetrahedron), you can now render the triangles using any 3D graphics library (like `matplotlib` in Python for basic visualization, or more advanced libraries like `Three.js` or `OpenGL`).
Here’s a simple example that projects the 4D points into 3D by dropping the 4th coordinate and then prints out the resulting 3D coordinates:
```python
import numpy as np

vertices_4d = np.array([
    [1, 0, 0, 0],
    [0, 1, 0, 0],
    [0, 0, 1, 0],
    [0, 0, 0, 1],
    [0, 1, 1, 1]
])

projected_vertices_3d = vertices_4d[:, :3]

print("Projected 3D vertices:")
print(projected_vertices_3d)
```

**Comment by maybeJosiah** [comment]
Are you suggesting trying to make a 3d shape out of 4d shapes because if so there is math for that out but 4d as a set of 3d spaces does not exactly convert so you may take a slice that goes through a view plane to a view point or some such thing using say a parametric equation of lines or in some cases just projecting line from end point on shape to view point and finding a set that actually is on view plane and to your set view point but honestly, if you want 4d to 3d straight any view point, I don't know and can't help, sorry maybe or maybe not, maybe.

**Reply by kijetesantakalu042** [reply]
For 3D graphics you can convert a 3d triangle into 3 2d lines. I was asking if anyone knew the math for converting a 4d tetrahedron into 4 3d triangles. I should maybe rephrase my question: How would you perspective project a 4D point into 3d space.
